client_unix/xs_client_unix.ml

```diff
@@ -301,11 +301,36 @@ functor
         with_mutex c.m (fun () -> Hashtbl.remove c.rid_to_wakeup rid);
         response hint request res unmarshal)
 
+    let directory_part h path offset =
+      rpc "directory_part" h
+        Request.(PathOp (path, Directory_part offset))
+        Unmarshal.raw
+
+    let rpc_dir hint h path =
+      try rpc hint h Request.(PathOp (path, Directory)) Unmarshal.list
+      with Error "E2BIG" ->
+        let rec read_part data generation =
+          let offset = String.length data in
+          let out = directory_part h path offset in
+          let i = String.index out '\000' in
+          let new_generation = String.sub out 0 i in
+          let new_data = String.sub out (i + 1) (String.length out - i - 1) in
+          (* Node's children changed, restart *)
+          if data <> "" && new_generation <> generation then read_part "" ""
+          else
+            let data = data ^ new_data in
```

Comment:
For efficiency it would be better to collect all strings in a list and do a `String.concat` or to use a `Buffer` value.

Reply:
Switched to using a `Buffer` in the fixup commit.